pkg/roachprod/roachprod.go

```diff
@@ -592,16 +592,29 @@ func Stage(
 	}
 
 	os := "linux"
-	arch := "amd64"
-
 	if c.IsLocal() {
 		os = runtime.GOOS
-		arch = runtime.GOARCH
 	}
+
+	// Default arch to the CPU architecture of the cluster's firt VM
+	// Architecture is determined using:
+	// - AWS & GCP: provider's cloud APIs (a.k.a machine-type describe)
+	// - Azure: hardcoded ARM64 machine-types
+	// - Local: runtime.GOARCH
+	detectedArch, err := c.GetCPUArchitecture(l)
+	if err != nil {
+		l.Printf("WARN: unable to detect remote architecture: %s", err)
+	}
+	arch := string(detectedArch)
```

Comment:
It's actually a really good point.

I tried as much as possible not to introduce a breaking change, so the (now optional) `--arch` arg keeps priority even in case of `ArchUnkwnow`. If we error out, this will break the "I know what I'm doing, let me force the architecture" behavior, so I would lean towards printing a big fat and ugly warn...

Reply:
It's also an anti-pattern... using result when `err` is non-nil. I believe we have a linter for that; wondering why it wasn't flagged?

Reply:
I'm trying to think if we need `ArchUnknown` anymore? I think returning an error already implies that we couldn't figure out the arch, except now we have some spots where we return `ArchUnknown` but no error. Just seems more prone to error if you ever need to check the result you now need to check both `err != nil` and `arch != ArchUnknown`

Reply:
> I'm trying to think if we need ArchUnknown anymore?

It's mostly for `func ParseArch(s string) CPUArch`. E.g., an error getting the architecture during `List` shouldn't fail it; instead we'd just display `unknown`. In other places like `attachToExistingCluster`, it should probably be still handled as a warning instead of a fatal.